backupccl: disallow restore of backups older than the minimum supported binary version

adityamaru opened the pull request:
As outlined in https://www.cockroachlabs.com/docs/v22.2/restoring-backups-across-versions.html,
we do not support restoring backups older than one major version. This
patch enforces this in restore code by ensuring that the manifest's version is greater
than equal to the minimum supported binary version of the cluster.

Release note (sql change): Backups older than one major version
from the cluster's current active version are not restoreable as per
the policy in https://www.cockroachlabs.com/docs/v22.2/restoring-backups-across-versions.html.

Fixes: #94295

adityamaru committed:
backupccl: disallow restore of backups older than the minimum supported binary version

As outlined in https://www.cockroachlabs.com/docs/v22.2/restoring-backups-across-versions.html,
we do not support restoring backups older than one major version. This
patch enforces this in restore code by ensuring that the manifest's version is greater
than equal to the minimum supported binary version of the cluster.

Release note (sql change): Backups older than one major version
from the cluster's current active version are not restoreable as per
the policy in https://www.cockroachlabs.com/docs/v22.2/restoring-backups-across-versions.html.

Fixes: #94295

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/93804)
<!-- Reviewable:end -->

adityamaru commented:
Lots of tests are going to break, I want CI to tell me what fixtures I need to regenerate.

adityamaru committed:
backupccl: remove interleaved tables test

Interleaved tables were removed >1 major version in
the past and so a backup restoreable in 23.1 cannot contain
interleaved tables.

Release note: None

adityamaru committed:
backupccl: remove exceptional backups

The xDBRef and duplicate-db exceptional backups
can no longer be restored since they were generated
>1 major version in the past. This change does add
a test inspired by the duplicate-db test to expand
coverage.

Release note: None

adityamaru committed:
backupccl: remove fk revision history test

This change removes the test that was added as part of
were correctly upgraded on restore to 20.x+ versions.
Backups taken one major release ago will no longer have
these old style fks and so this test serves better as a
regular test on this branch.

Release note: None

adityamaru committed:
backupccl: update `cluster` fixture and remove 20.1 zcfg fixutre

This change removes old cluster restore fixtures, and replaces them
with a cluster backup from 22.2.0. It also removes a 20.1 fixture that
tested restore of a zoneconfig privilege that was overwritten as the usage
privilege in 20.2+ versions. This test has been migrated to a regular
test instead.

Release note: None

adityamaru committed:
backupccl: removes old MR backup fixture

This change removes an old MR backup that was being
restored and tested against. There were no instructions
on how to regenerate the fixture in #69725 so it is being
deleted.

Release note: None

adityamaru committed:
backupccl: remove public schema remap fixtures

Prior to 22.2 the public schema was not backed by
a descriptor. These fixtures tested that a restore
of a pre 22.2 backup would remap the synthetic public
schema to a descriptor backed public schema. Since the
only restoreable backups on this branch will be from
22.2 onwards, we should never see a synthetic public schema
in a restoreable backup.

Release note: None

adityamaru committed:
backupccl: delete TestRestoreOldBackupMissingOfflineIndexes

The backup fixtures in this test were generated on 20.2 where
https://github.com/cockroachdb/cockroach/issues/62564 existed.
Given this was >1 major version in the past we do not expect to
see a reoccurrence of the bug in 22.2+ backups that are restoreable
in this release.

Release note: None

adityamaru committed:
backupccl: remove dropped schema 20.2.7 fixture

This fixture captured a bug in 20.2.7 that has since
been fixed. This change removes the fixture and moves it
to a regular regression test.

Release note: None

adityamaru committed:
backupccl: delete exports without interleaved

These fixtures were generate on older version of
Cockroach that are no longer restoreable. The test
has been migrated to a regular datadriven test.

Release note: None

adityamaru committed:
backupccl: regenerate sequences fixture on 22.2.0

This change regenerate the sequences backup fixture on 22.2.0.
While there have been no changes to the representation of sequences
between 22.2 and 23.1 it seemed useful to keep a cross version
backup test of this nature.

Release note: None

adityamaru committed:
backupccl: delete fixtures for users with no ID

All users in 22.2+ clusters will have IDs and so
will the backups taken on these clusters. Thus, there
is no need to keep around the cross version test
that tests the custom restore func that injects a
user ID.

Release note: None

adityamaru committed:
backupccl: regenerate fixtures for restoring mid schema change

This change regenerates the fixtures for testing
restores of backups that have been taken before or after
the backfill portion of a schemachange. Note these tests
rely on the old schemachanger and do not increase our test
coverage with the new declarative schema changer.

Release note: None

adityamaru committed:
backupccl: delete missing-import backup test

This test was testing a bug that was discovered during
backups in pre-22.2 versions https://github.com/cockroachdb/cockroach/issues/88042.
Since this bug does not exist on 22.2+ versions
this test will not encounter the expected corruption
on backups that are restoreable in this cluster.

Release note: None

adityamaru commented:
I have one TODO to resolve https://github.com/cockroachdb/cockroach/pull/93804/commits/743d5e1df8a1d96f08ee2d4f17344732dfffa1f3#diff-87e52e674b19b4c9ef048ee88b0a6b5feec243b869bb266a5551826f6997b0c8R72 and ensure our restore roachtests are up to date with this policy, but this should be ready for a look.

I will squash all the commits once I get an LGTM!

adityamaru committed:
backupccl: add env variable to disable version checking

This change adds an env variable that will be set by the
job running restore roachtests until we have updated all their
fixtures to be 22.2.0+.

Release note: None

stevendanna commented on pkg/ccl/backupccl/restore_planning.go:
```diff
@@ -1504,14 +1506,44 @@ func doRestorePlan(
-			}
+		v := mainBackupManifests[i].ClusterVersion
+		if v.Major == 0 {
+			return errors.WithHint(errors.Newf("the backup is from a version older than our "+
+				"minimum restoreable version %s", minimumRestorableVersion),
+				"refer to our documentation about restoring across versions: https://www.cockroachlabs.com/docs/v22.2/restoring-backups-across-versions.html")
```
Does the v.Major == 0 case imply we never recorded a version in the manifest?  If so, does it definitely need to happen before the currentVersion.Less check since current version will always be greater? that would help keep the two related checks next to each other.

stevendanna reviewed (commented):
I didn't read all of the test code updates, but overall this seems reasonable if we still want to enforce this in code. Left one small question.

adityamaru commented:
Closing in favour of #98597.
